pkg/agent/agent.go

```diff
@@ -900,47 +900,60 @@ func (i *Initializer) initNodeLocalConfig() error {
 		return nil
 	}
 
-	// Parse all PodCIDRs first, so that we can support IPv4/IPv6 dual-stack configurations.
-	if node.Spec.PodCIDRs != nil {
-		for _, podCIDR := range node.Spec.PodCIDRs {
-			_, localSubnet, err := net.ParseCIDR(podCIDR)
-			if err != nil {
-				klog.ErrorS(err, "Failed to parse subnet from Pod CIDR string", "CIDR", podCIDR)
-				return err
-			}
-			if localSubnet.IP.To4() != nil {
-				if i.nodeConfig.PodIPv4CIDR != nil {
-					klog.InfoS("One IPv4 PodCIDR is already configured on this Node, ignoring the IPv4 Subnet CIDR", "subnet", localSubnet)
+	err = wait.PollImmediate(10*time.Second, 30*time.Second, func() (bool, error) {
+		node, err = i.client.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to get Node with name %s from K8s: %w", nodeName, err)
+		}
+		// Parse all PodCIDRs first, so that we can support IPv4/IPv6 dual-stack configurations.
+		if node.Spec.PodCIDRs != nil {
+			for _, podCIDR := range node.Spec.PodCIDRs {
+				_, localSubnet, err := net.ParseCIDR(podCIDR)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse subnet from Pod CIDR string", "CIDR", podCIDR)
+					return false, err
+				}
+				if localSubnet.IP.To4() != nil {
+					if i.nodeConfig.PodIPv4CIDR != nil {
+						klog.InfoS("One IPv4 PodCIDR is already configured on this Node, ignoring the IPv4 Subnet CIDR", "subnet", localSubnet)
+					} else {
+						i.nodeConfig.PodIPv4CIDR = localSubnet
+						klog.V(2).InfoS("Configured IPv4 Subnet CIDR on this Node", "subnet", localSubnet)
+					}
+					continue
+				}
+				if i.nodeConfig.PodIPv6CIDR != nil {
+					klog.InfoS("One IPv6 PodCIDR is already configured on this Node, ignoring the IPv6 Subnet CIDR", "subnet", localSubnet)
 				} else {
-					i.nodeConfig.PodIPv4CIDR = localSubnet
-					klog.V(2).InfoS("Configured IPv4 Subnet CIDR on this Node", "subnet", localSubnet)
+					i.nodeConfig.PodIPv6CIDR = localSubnet
+					klog.V(2).InfoS("Configured IPv6 Subnet CIDR on this Node", "subnet", localSubnet)
 				}
-				continue
 			}
-			if i.nodeConfig.PodIPv6CIDR != nil {
-				klog.InfoS("One IPv6 PodCIDR is already configured on this Node, ignoring the IPv6 Subnet CIDR", "subnet", localSubnet)
+			return true, nil
+		}
+		if node.Spec.PodCIDR != "" {
+			_, localSubnet, err := net.ParseCIDR(node.Spec.PodCIDR)
+			if err != nil {
+				return false, fmt.Errorf("failed to parse subnet from CIDR string %s: %w", node.Spec.PodCIDR, err)
+			}
+			if localSubnet.IP.To4() != nil {
+				i.nodeConfig.PodIPv4CIDR = localSubnet
```

Comment:
Is there any other place to initialize `i.nodeConfig.PodIPv4CIDR`? why we didn't check `i.nodeConfig.PodIPv4CIDR != nil` here?

Reply:
I'm not sure that I understand the question but I'll try:
The node spec has two different vars which hold the Pod CIDR: The legacy PodCIDR which can hold a single CIDR and PodCIDRs which is a list. First the code tries PodCIDRs, yet as our agent uses only one IP4 and one IP6 CIDRs it checks whether the i.nodeConfig field is populated already.
If PodCIDRs isn't populated the code used PodCIDR which contains only a single CIDR - therefore i.nodeConfig isn't populated for sure.